### PR TITLE
Bind mouse buttons

### PR DESCRIPTION
**Pull request by samcaulfield**
These patches accomplish the following:

1. Create two new actions: dig and place, for digging and placing blocks. These actions can be bound to any key just like the other player actions e.g. sneak, jump. This involves removing hardcoded associations between the mouse and these actions from game.cpp.
2. Allow the mouse buttons to be bound to any action in the GUI just like any other button.

### CODE REVIEW

**Comment by jastevenson303** [comment]
Looking forward to testing this as soon as possible.  Thanks.

File: src/game.cpp [hunk 1]
```diff
@@ -3951,11 +3953,11 @@ void Game::handlePointingAtObject(const PointedThing &pointed, const ItemStack &
 			runData.object_hit_delay_timer = object_hit_delay;
 		}
 
-		if (getLeftClicked())
+		if (isKeyDown(KeyType::DIG))
 			do_punch = true;
 
 		if (do_punch) {
-			infostream << "Left-clicked object" << std::endl;
+			infostream << "Dug object" << std::endl;
```
Comment (SmallJoker) [inline src/game.cpp]: `Punched object`

File: src/game.cpp [hunk 2]
```diff
@@ -3978,8 +3980,8 @@ void Game::handlePointingAtObject(const PointedThing &pointed, const ItemStack &
 			if (!disable_send)
 				client->interact(0, pointed);
 		}
-	} else if (getRightClicked()) {
-		infostream << "Right-clicked object" << std::endl;
+	} else if (wasKeyDown(KeyType::PLACE)) {
+		infostream << "Dug object" << std::endl;
```
Comment (SmallJoker) [inline src/game.cpp]: `Punched object`

File: src/game.cpp (hunk 1 above)
Comment (samcaulfield) [reply src/game.cpp]: I've changed the string and renamed left_punch to just punch

File: src/game.cpp (hunk 2 above)
Comment (samcaulfield) [reply src/game.cpp]: I changed the string to "Pressed place button while looking at object". More verbose but less ambiguous

**Comment by samcaulfield** [comment]
There's now a minor UX issue on the key binding screen that I'd like to fix. When the button to set the key binding for an action is clicked, if the desired binding is the left mouse button, then the GUI will eat the click and won't set the binding unless the click occurs somewhere else other than the button. This only occurs for LMB.

**Comment by samcaulfield** [comment]
I have fixed the above issue. The workaround is not ideal, but hopefully I have documented it well enough so that someone more familiar with Irrlicht's GUI system could come up with a more elegant solution in the future.

That pretty much completes this feature.

**Comment by jastevenson303** [comment]
Works beautifully, so far so good.  I only updated my client, changed keys, connected to server and it worked great.  Thank you very much!

**Comment by jastevenson303** [comment]
@samcaulfield Can you test eating an apple?  I seem to be eating a multitude of items on a single button press.

**Comment by jastevenson303** [comment]
I'm just throwing around ideas, but does it matter that safe_dig_and_place is checked on place, but not dig?  https://github.com/minetest/minetest/pull/6537/files#diff-18513665750ef5adf42b5ec29e14162eR3691

I didn't know about that setting prior, but setting it here didn't help me.  As an aside, I always thought repeat_rightclick_time was for the inventory formspecs to right-click spread items.  Didn't know it was for placing nodes, or even that that was adjustable.

I thought it seemed as though placing nodes was happening more quickly with the key depressed, now.  I'm still using this PR, with space and RMB swapped, and don't want to go back.  :D

File: builtin/settingtypes.txt [hunk 3]
```diff
@@ -96,8 +96,8 @@ doubletap_jump (Double tap jump for fly) bool false
 #    If disabled, "special" key is used to fly fast if both fly and fast mode are enabled.
 always_fly_fast (Always fly and fast) bool true
 
-#    The time in seconds it takes between repeated right clicks when holding the right mouse button.
-repeat_rightclick_time (Rightclick repetition interval) float 0.25
+#    The time in seconds it takes between repeated block placements when holding the place block button.
+repeat_place_time (Rightclick repetition interval) float 0.25
```
Comment (jastevenson303) [inline builtin/settingtypes.txt]: Still mentions rightclick, "Rightclick repetition interval" ... `Place key repetition interval`?

**Comment by samcaulfield** [comment]
Thanks for finding that issue. I've found the root cause of it. I had replaced a getClicked with an isButtonDown. I'll look into emulating clicks for buttons generally.

**Comment by jastevenson303** [comment]
Alright, sounds great!

I wanted to add that I just noticed - in third person camera mode - that Sam's digging animation is still hard-coded to RMB.

**Comment by jastevenson303** [comment]
@samcaulfield I changed `isKeyDown` to `wasKeyDown` for a usable item, and that seemed to work.  I'll test more and report back.  (Place button pointing at air still spams usable item in console, however.)
```
diff --git a/src/game.cpp b/src/game.cpp
index e9602891..84c69f64 100644
--- a/src/game.cpp
+++ b/src/game.cpp
@@ -3700,7 +3700,7 @@ void Game::processPlayerInteraction(f32 dtime, bool show_hud, bool show_debug)
                runData.repeat_place_timer = 0;
 
        if (playeritem_def.usable && isKeyDown(KeyType::DIG)) {
-               if (isKeyDown(KeyType::DIG) && (!client->moddingEnabled()
+               if (wasKeyDown(KeyType::DIG) && (!client->moddingEnabled()
                                || !client->getScript()->on_item_use(playeritem, pointed)))
                        client->interact(4, pointed);
        } else if (pointed.type == POINTEDTHING_NODE) {
```

**Comment by jastevenson303** [comment]
- [x] Fix dig/use button repeating on single click https://github.com/jastevenson303/minetest/commit/837dad528d48cb42789c6832dc36ba3569749e01
- [ ] Fix place/usable button repeating on activate (pointing at nothing)
- [ ] Fix get_player_control().RMB (visible in 3rd person when jumping, hand swing animation plays)

I don't know about get_player_control().LMB, didn't test or see.  @paramat @rubenwardy I didn't know who to highlight after @samcaulfield, although @lhofhansl gave a thumbs up.  I fixed the repeat on use with a strong hint from author, but not a clue about the other bugs.

**Comment by samcaulfield** [comment]
Apologies for leaving this PR for a while, had stuff going on IRL which pushed this to the back of my mind. Last time I worked on this I had dug quite deep into the input handling code, all the way into Irrlicht's input handling sitting underneath Minetest. IIRC, I found that fixing these issues would probably require a lot of work because of the way Irrlicht is designed, similar to the author of the previous PR which attempted similar changes.

Maybe I'm wrong though. I'll have to re-accquaint myself with these changes to remind myself.

**Comment by jastevenson303** [comment]
It's just a little heart breaking, and I've made something of a mistake in
getting used to right-mouse to jump again.  I've been using this for all
this time now, and it works beautifully -- to my mind, they seem like small
bugs, although I surely wouldn't know.  I figure, if
get_player_control().sneak knows I bound this to Z and not the default
Shift, then why wouldn't get_player_control().RMB?  I assumed fixing the
item place auto-repeat bug would be similar to the item usable fix.  I wish
you the best in any case, and thanks again.  It's been a treat!  :^)

**Comment by rubenwardy** [comment]
alternative implementation: https://github.com/minetest/minetest/pull/5699

**Comment by paramat** [comment]
@samcaulfield will you be able to work on this again soonish? if it's a long way off we could temporarily close if that's ok with you.

**Comment by samcaulfield** [comment]
Hey, I'll take another look at this today.

File: builtin/settingtypes.txt (hunk 3 above)
Comment (samcaulfield) [reply builtin/settingtypes.txt]: Good catch, thanks

**Comment by samcaulfield** [comment]
I have done a lot of reading into the engine's input handling internals along with some cleanup. I believe most of the bugs we're seeing are happening because of the move away from functions like getLeftClicked() to isKeyDown(), i.e. something was lost in translation moving from the old mouseclick system to the new system that treats everything as a key press. It's probably why we're seeing stuff like actions repeating rapidly when they shouldn't be. I'll work on ironing that out over the next few weeks.

I think a good way to do input systems is to treat each unique action a person can do with their input device as an abstract "Action" and to not care if it was a mouse move, mouse click, key press, joystick input etc. A good example of this is Arma 3's keybinding system where you can bind anything to anything, e.g. you can bind "Mouse move +X axis" to fire a gun etc. Very versatile. This is a bit of a tangent but I think it helps to explain where I hope to see this PR leading to eventually.

@jastevenson303 I'm a bit rusty with Minetest/craft-style user interface & controls (I started working on this PR almost immediately after trying out Minetest. I always rebind leftmouse to F key), so your testing has been very helpful.

**Comment by samcaulfield** [comment]
That's exactly how I was thinking of approaching it. Minetest should see an InputHandler that completely abstracts away the details of the underlying input system. You would theoretically be able to swap out the input library completely and no Minetest application-level code would have to be changed. In fact, Minetest probably shouldn't even know what input device is actually being used, so there would be no references to mouse/joystick/touchscreen etc. anywhere in Minetest.

I know such a large change is outside the scope of this PR, but I'll keep it in mind as I work on this because it's a good foundation. So far, we've pretty much removed all of the mouse hardcoding in the engine.

Depending on how successful I can get with this I'd be interested in looking at the whole input system later on.

**Comment by paramat** [comment]
Note we have 2 ideas for using middle mouse button, so a binding for that too will be useful, especially as some mouse-equivalent-devices have no middle button and poor workarounds for replicating middle click.

**Comment by jastevenson303** [comment]
All prior bugs are fixed in this new version @samcaulfield thank you once again!

:+1:

**Comment by samcaulfield** [comment]
The only remaining issue I can see is that holding down the dig key while holding a consumable (e.g. apple) causes the animation to repeat while the button is held (apple is only eaten once though, correctly). It should only play athe animation once.

I haven't done any testing on Android/joystick/touchscreens though.

**Comment by jastevenson303** [comment]
I can try it on other devices.  @paramat has a trackball...
Is that the typical previous behavior with the animation or something newly
specific to this PR?

**Comment by samcaulfield** [comment]
That behaviour is new to this PR, most likely a bug introduced in game.cpp when some isClicked logic was replaced with isPressedDown

**Comment by jastevenson303** [comment]
There seems to be a conflict now, and when I resolve it manually it won't let me connect to the straight server, giving an error:
```
2018-01-08 02:19:45: ERROR[Main]: A serialization error occurred:
2018-01-08 02:19:45: ERROR[Main]: unsupported NodeBox version
2018-01-08 02:19:45: ERROR[Main]: The server is probably  running a different version of Minetest.
```

**Comment by jastevenson303** [comment]
It's now not possible for me to resolve the conflict, it's really big!  Before, it was like one line, but not anymore.  @samcaulfield?

**Comment by samcaulfield** [comment]
It seems there was a refactor to the game and input handler which is good too see. It shouldn't be too difficult to merge. I'll get on that today.

**Comment by Ferk** [comment]
> holding down the dig key while holding a consumable (e.g. apple) causes the animation to repeat while the button is held 

Does it make any difference to this if you are holding a stack with multiple apples vs only 1 apple?

It could be it's related to timing issues when switching to the hand, like it happened in #1938

**Comment by jastevenson303** [comment]
That was an old issue since fixed.

**Comment by jastevenson303** [comment]
This should probably just be closed, if it's been a year and there's been at least one refactor, and little interest.

**Comment by Wuzzy2** [comment]
Wait, what? So this PR fails only because of a missing refactor? OMG.

**Comment by paramat** [comment]
If only rebase is blocking no need to close. This needs review and some support before we ask for a major rebase. @rubenwardy any thoughts?

**Comment by samcaulfield** [comment]
Hi, sorry for forgetting about this PR.

Unfortunately I don't have time to work on this at the moment. Besides, I think it would be much more time-efficient if someone more familiar than I with Minetest's input system picked this up to fix the last couple of (known) bugs.

I tried to make the changeset for this PR as simple as possible so hopefully it would be easy for someone else to pick up and rebase.

**Comment by paramat** [comment]
No problem and thanks for letting us know.

**Comment by ClobberXD** [comment]
I'm willing to adopt this PR.

> 3. Needs Lua API backwards compatibility code (renaming the fields breaks many mods)

@SmallJoker could you please elaborate on this? What exactly needs to be done to fix this compat breakage?

**Comment by paramat** [comment]
Replaced by #7924

**Pull request closed without merging**